packages/grpc-js-core/src/channel.ts

```diff
@@ -62,49 +72,97 @@ export interface Channel extends EventEmitter {
 
 export class Http2Channel extends EventEmitter implements Channel {
   private connectivityState: ConnectivityState = ConnectivityState.IDLE;
-  private idleTimerId: NodeJS.Timer|null = null;
   /* For now, we have up to one subchannel, which will exist as long as we are
    * connecting or trying to connect */
   private subChannel: http2.ClientHttp2Session|null;
   private filterStackFactory: FilterStackFactory;
 
-  private transitionToState(newState: ConnectivityState): void {
-    if (newState !== this.connectivityState) {
+  private idleTimerId: NodeJS.Timer|null = null;
+
+  private subChannelConnectCallback: ()=>void = () => {};
+  private subChannelCloseCallback: ()=>void = () => {};
+
+  private backoffTimerId: NodeJS.Timer|null = null;
+  private currentBackoff: number = INITIAL_BACKOFF_MS;
+  private currentBackoffDeadline: Date;
+
+  private handleStateChange(oldState: ConnectivityState, newState: ConnectivityState): void {
+    let now: Date = new Date();
+    switch(newState) {
+    case ConnectivityState.CONNECTING:
+      if (oldState === ConnectivityState.IDLE) {
+        this.currentBackoff = INITIAL_BACKOFF_MS;
+        this.currentBackoffDeadline = new Date(now.getTime() + INITIAL_BACKOFF_MS);
+      } else if (oldState == ConnectivityState.TRANSIENT_FAILURE) {
+        this.currentBackoff = Math.min(this.currentBackoff * BACKOFF_MULTIPLIER, MAX_BACKOFF_MS);
+        let jitterMagnitude: number = BACKOFF_JITTER * this.currentBackoff;
+        this.currentBackoffDeadline = new Date(now.getTime() + this.currentBackoff + uniformRandom(-jitterMagnitude, jitterMagnitude));
+      }
+      this.startConnecting();
+      break;
+    case ConnectivityState.READY:
+      this.emit('connect');
+      break;
+    case ConnectivityState.TRANSIENT_FAILURE:
+      this.subChannel = null;
+      this.backoffTimerId = setTimeout(() => {
```

Comment:
`this.backoffTimerId` is never used. Should it be used to cancel the timeout on shutdown?

Reply:
yes, I've added that.